### PR TITLE
Don't loose signatures in template let binding 

### PR DESCRIPTION
**Pull request by nickchapman-da**
Don't loose type signatures in template let binding

- addressing issue: https://github.com/digital-asset/daml/issues/7902
- ghc patch is here: https://github.com/digital-asset/ghc/pull/56

A new testcase (from the issue) demonstrates that the type signatures are now checked, but it also demonstrates that we get 5 error messages for the same type error.

This is a more general issue, caused by the way we duplicate code in the desugaring of templates-lets.  The issue is not limited to type signatures, it applies to all warnings produced by let bindings.  We can’t avoid recomputing let bindings in each choice but we could factor out the whole logic including the type signatures for computing the let bindings to produce the warning only once. We should do this in a future PR. There is a separate issue for this problem: https://github.com/digital-asset/daml/issues/2175

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: stackage_snapshot.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
```
Comment (cocreature) [inline stackage_snapshot.json:1]: Also needs an update on Windows.

**Review by cocreature — APPROVED** [review]
Thanks! Please bump the commit to the one after merging https://github.com/digital-asset/ghc/pull/56#pullrequestreview-543791212

File: compiler/damlc/tests/daml-test-files/TemplateLetWithSignatures.daml [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+-- All rights reserved.
+
+module TemplateLetWithSignatures where
+
+-- This example should fail to type check.
+-- There was a bug reported: https://github.com/digital-asset/daml/issues/7902
+-- which indicated that the signature was being ignored.
+-- It is now not ignored, and the example does indeed fail to type check.
+
+-- TODO: only report the type error once, instead of 5 times!
```
Comment (hurryabit) [inline compiler/damlc/tests/daml-test-files/TemplateLetWithSignatures.daml:11]: The more choices you add, the more errors you get.

Do you have an idea how to fix this?

**Review by hurryabit — APPROVED** [review]
Very nice. Thank you.

File: compiler/damlc/tests/daml-test-files/TemplateLetWithSignatures.daml (hunk 2 above)
Comment (nickchapman-da) [reply compiler/damlc/tests/daml-test-files/TemplateLetWithSignatures.daml:11]: I guess a new top level function which is passed this and returns a tuple of the values, which get bound to the vars in each place.

File: stackage_snapshot.json (hunk 1 above)
Comment (nickchapman-da) [reply stackage_snapshot.json:1]: bump: done
windows: right....

Comment (nickchapman-da) [reply stackage_snapshot.json:1]: Using a windows ad-hoc machine,  updated: `stackage_snapshot_windows.json`
First time I've managed to get this to work!
Thanks @garyverhaegen-da for walking me through the setup.
I will record the steps in the working on `ghc-lib` document...

File: compiler/damlc/tests/daml-test-files/TemplateLetWithSignatures.daml (hunk 2 above)
Comment (nickchapman-da) [reply compiler/damlc/tests/daml-test-files/TemplateLetWithSignatures.daml:11]: This should get fixed as when issue  https://github.com/digital-asset/daml/issues/2175 is addressed.

**Pull request merged**